### ISSUE TITLE
Moving a marker programmatically destroys and recreates it

### ISSUE THREAD

**Issue by cyrilchapon** [tombatossals/angular-leaflet-directive#789]
Moving a marker (updating it's lat or lng property), doesn't **move** it, but **destroys and recreates** it on the new location, closing any popup in the process.

```
angular.extend($scope, {
    ...,
    markers: {}
  }
});
```

...first adding some markers, then calling 

```
var updateMarkerFromBean = function(marker, bean) {
  marker.lat = bean.lastLocation.geopoint.coordinates[1];
  marker.lng = bean.lastLocation.geopoint.coordinates[0];
}
```

Is that intended to work this way ? Is there a way to realy move it ?

**Comment by lambdatoast** [comment]
I have this issue as well. 

There appears to be some logic in markers.js that diffs markers, and destroys them whenever they change.

**Comment by nmccready** [comment]
Change the model, or get a copy of the original model and then create the marker manually.

**Comment by nmccready** [comment]
This is working as designed, so I do not see the problem.

**Comment by cyrilchapon** [comment]
> I do not see the problem.

Here it is:

> closing any popup in the process.

Any real life situation with auto updating locations, through polling or sockets, will change the marker lng and lat, closing opened popup.
-----

> This is working as designed

Well, thus the problem is the way its designed. My point here is not "it's not working as designed", my point is "changing any marker object property implying it to destroy and recreate is strange"

**Comment by lambdatoast** [comment]
> Change the model, or get a copy of the original model and then create the marker manually.

That is, don't use the markers functionality of the angular-leaflet-directive. OK.

> This is working as designed, so I do not see the problem.

So it is designed for maps which have markers whose contents never change. This should be disclaimed in the docs, as Angular is often used for interactive applications with time-varying values.

**Comment by nmccready** [comment]
No it is designed for the markers to change so obviously there is an issue. I was tired when I read this and providing an example like here #805 makes it a whole lot easier to figure out. If its broke then you are just as capable of digging in and trying to figure it out as well.

**Comment by nmccready** [comment]
I'm marking this as a dupe as #805 appears to be the same issue.

**Comment by nmccready** [comment]
Ok, in actually sitting down with a clear head; (been a busy past few days) I get your issue now. This is basically the same problem as the marker "flash" in agm. So originally when I got involved in this directive all markers would be erased and then drawn regardless of the update. I changed it to check if anything different happened. If that marker did change then it would be erased. The enhancement would be to update the logic to just move the marker instead of blowing it away. The catch is that in the _addMarkers portion it would also need to make sure to not duplicate the marker as well. Once this is done I think everyone will be satisfied.

**Comment by cgat** [comment]
Here's my crack at it https://github.com/cgat/angular-leaflet-directive/tree/updating-markers

Lot's of the logic for updating the marker is already in addMarkerWatchers. I pulled this out and made it and made it an updateMaker. Please have a look and tell me what you think.

**Comment by nmccready** [comment]
Made comments on your commit.

**Comment by cgat** [comment]
Sorry, that was a bad commit on my part. I thought I was pushing a commit I hadn't made. Look at the latest one. This doesn't solve the popup issue, which is a different can of worms.

**Comment by nmccready** [comment]
More comments mainly the renaming of the function to updateMarkers should be in a follow up PR as it is making it way hard to review. @cgat

**Comment by cgat** [comment]
@nmccready Yeah, that PR is hard to read, especially with the messed up commits in between.

I've made a new PR branch, which now makes it easier to see the difference between master and my branch. I've also made a bunch of comments to help guide you along. It's not as complicated as it looks. 

See the commit here: https://github.com/cgat/angular-leaflet-directive/commit/f1309eed5a721083f9d314c7b850316e3fea185f

or the branch here: https://github.com/cgat/angular-leaflet-directive/tree/updating-markers2

I suggest you start by examining addMarkerWatchers (the original), and see how I modified it. All I really did was pull out the watch callback into an accessible function that could be shared, then call that in _addMarker in the directive. 

Cheers
